src/Storages/MergeTree/MergeTreeIndexGranularityInfo.cpp

```diff
@@ -29,14 +29,14 @@ std::optional<std::string> MergeTreeIndexGranularityInfo::getMarksExtensionFromF
     return {};
 }
 
-MergeTreeIndexGranularityInfo::MergeTreeIndexGranularityInfo(const MergeTreeData & storage, MergeTreeDataPartType type_)
+MergeTreeIndexGranularityInfo::MergeTreeIndexGranularityInfo(const MergeTreeData & storage, MergeTreeDataPartType type_, bool is_virtual)
     : type(type_)
 {
     const auto storage_settings = storage.getSettings();
     fixed_index_granularity = storage_settings->index_granularity;
 
     /// Granularity is fixed
-    if (!storage.canUseAdaptiveGranularity())
+    if (!storage.canUseAdaptiveGranularity() && !is_virtual)
```

Comment:
Here, `is_virtual` is a strange name. What `virtual` index granularity is?
Maybe something like `firce_disable_adaptive_granularity`, and a comment why it is needed.

Also, for me it is not clear - what kind of error we want to avoid?